app/assets/javascripts/sitewide/drag_n_drop.js.erb

```diff
@@ -0,0 +1,350 @@
+(function(global) {
+  'use strict';
+
+  // Module to handle file uploading in Results
+  global.DragNDropSteps = (function() {
+    var droppedFiles = [];
+    var filesValid = true;
+    var totalSize = 0;
+
+    function init(files, action) {
+      for(var i = 0; i < files.length; i++) {
+        droppedFiles.push(files[i]);
+      }
+      if(action === 'select') {
+        listItems();
+      }
+    }
+
+    // return the status of files if they are ready to submit
+    function filesStatus() {
+      return filesValid;
+    }
+
+    // loops through a list of files and display each file in a separate panel
+    function listItems() {
+      droppedFiles = droppedFiles.filter(Boolean);
+      $('.panel-step-attachment-new').remove();
+      _dragNdropAssetsOff();
+      for(var i = 0; i < droppedFiles.length; i++) {
+        $('#new-step-assets')
+          .append(_uploadedAseetPreview(droppedFiles[i], i))
+          .promise()
+          .done(function() {
+            _removeItemHandler(i);
+          });
+      }
+      _validateTotalSize();
+      dragNdropAssetsInit('steps');
+    }
+
+    // appent the files to the form before submit
+    function appendFilesToForm(ev) {
+      var regex = /step\[assets_attributes\]\[[0-9]*\]\[id\]/;
+      var prevEls = $('input').filter(function() {
+        return this.name.match(regex);
+      });
+      droppedFiles = droppedFiles.filter(Boolean);
+      var fd = new FormData($(ev.target).closest('form').get(0));
+      for(var i = 0; i < droppedFiles.length; i++) {
+        var index = i + prevEls.length;
+        var name = 'step[assets_attributes][' + index + '][file]';
+        fd.append(name, droppedFiles[i]);
+      }
+      droppedFiles = [];
+      filesValid = true;
+      totalSize = 0;
+      _dragNdropAssetsOff();
+      return fd;
+    }
+
+    function _validateFilesSize(file) {
+      if((file.size/1048576) > <%= Constants::FILE_MAX_SIZE_MB %> && filesValid) {
+        return "<p><%= I18n.t 'general.file.size_exceeded', file_size: Constants::FILE_MAX_SIZE_MB %></p>";
+      }
+      totalSize += parseInt(file.size/1048576);
+      return '';
+    }
+
+    function _validateTotalSize() {
+      if(totalSize > <%= Constants::FILE_MAX_SIZE_MB %>) {
+        filesValid = false;
+        $.each($('.panel-step-attachment-new'), function() {
+          $(this)
+            .find('.panel-body')
+            .append("<p class='dnd-error'><%= I18n.t('general.file.total_size', size: Constants::FILE_MAX_SIZE_MB) %></p>");
+        });
+      } else {
+        $('.dnd-error').remove();
+        filesValid = true;
+      }
+    }
+
+    function _uploadedAseetPreview(asset, i) {
```

Comment:
Like I said there are some duplications but here we have separate modules and this methods is private so for now I will leave it in like it is.